Fix torch::deploy workaround again.

kumpera opened the pull request:
There are two issues, can't init Library unconditionally and the torch::deploy
test is wrong. Let's breakdown:

1) Can't init Library unconditionally.

We create a "DEF" Library and that never worked under torch::deploy, to the point
that it stubs out torch._meta_registrations to avoid triggering it.

We still need to create it at the module level or the whole thing breaks. Moving
the creation under the if body fixes this.

2) torch::deploy test is wrong

The support code to the original probing method was removed in August 2022.
We're now using the fact that torch._meta_registrations is stubbed out instead.

wconstab commented on torch/__init__.py:
```diff
@@ -1650,3 +1650,7 @@ def _sparse_coo_tensor_unsafe(*args, **kwargs):
+
+# Register functional collectives
+import torch.distributed._functional_ops as _functional_ops
+del _functional_ops
```
im not sure what the point of this is, probably need a comment here

kumpera replied:
The point is ensuring no namespace pollution.
We want make sure that nobody can do `import torch._functional_ops as big_mistake`

wconstab replied:
i see, yea that makes sense. you're just doing this to trigger the init for deploy, iiuc

wconstab commented on torch/distributed/_functional_collectives.py:
```diff
@@ -64,6 +62,18 @@
+Add the op definition to _functional_ops.py::ops_defs, the format is the one torch.library expects.
+Add _{op} function to this file with matching signature. This is the op implementation.
+Add _{op}_meta function to this file with the meta kernel to be used by symbolic tracing.
+aAdd {op} high level version to this file. It's what users are expected to call.
```
typo

wconstab commented on torch/distributed/_functional_ops.py:
```diff
@@ -0,0 +1,35 @@
+    "reduce_scatter_tensor(Tensor input, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor",
+]
+
+def _register_ops(resolve_ops):
```
why is resolve_ops passed here, it's not used?

kumpera replied:
Yeah, it's half done. Fixed :/

wconstab commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -79,7 +79,7 @@ c10::optional<OperatorHandle> Dispatcher::findOp(const OperatorName& overload_na
 void Dispatcher::waitForDef(const FunctionSchema& schema) {
   using namespace std::chrono_literals;
   std::unique_lock<std::mutex> lock(mutex_);
-  bool r = cond_var_.wait_for(lock, 2s, [&]{
```
oh, what is this?

kumpera replied:
This is the setup that enables multiple interpreters to handle op registration.
It requires waiting for ops to pop up with a given timeout.
I'll remove this change if my latest hack works. 🤞

wconstab commented on torch/distributed/_functional_ops.py:
```diff
@@ -0,0 +1,31 @@
+
+def _register_ops(resolve_ops):
+    import torch.distributed._functional_collectives as fun_col
+    my_module = fun_col
```
shouldn't fun_col only be imported in the `if torch.distributed.is_available() True` case?  It looks like you'll import it even if not available here.  I think you meant to delete those 2 lines above?

kumpera replied:
Import inside functions are only evaluated when the function is first called. That's why this works. :)

I removed this as I simplified the change even further.

kumpera committed:
Fix torch::deploy workaround again.

There are two issues, can't init Library unconditionally and the torch::deploy
test is wrong. Let's breakdown:

1) Can't init Library unconditionally.

We create a "DEF" Library and that never worked under torch::deploy, to the point
that it stubs out torch._meta_registrations to avoid triggering it.

We still need to create it at the module level or the whole thing breaks. Moving
the creation under the if body fixes this.

2) torch::deploy test is wrong

The support code to the original probing method was removed in August 2022.
We're now using the fact that torch._meta_registrations is stubbed out instead.

ezyang reviewed (approved):
k
